articles/generalized_interface_generation.md

```diff
@@ -0,0 +1,216 @@
+Thus, in the most general case, build system integration is necessary.
+
+However, integration with any given build system or build system generator couples a large portion of the interface pipeline to that set of tools, which then has to be ported over for projects that use other build systems or build system generators.
+Programs that can bridge across build systems and/or build system generators do exist, but these are rare and often limited to a subset of the functionality that is shared.
+
+By generating a build specification in a format designed to be simple yet general, an appropriate build system can be generated on consumption.
+
+##### Specification
+
+Interface type representation and type support build configuration CLIs are roughly equivalent but conceptually different, and thus kept separate.
+This allows them to deviate from each other in the future if need be.
+
+*For interface type representations*
+```sh
+rosidl_build_configure ((-o|--output-file) PATH)? ((-l|--language) IDENTIFIER(==VERSION)?)* ((-I|--include-path) PATH)* PATH+
+```
+
+*For interface type supports*
+```sh
+rosidl_typesupport_build_configure ((-o|--output-file) PATH)? ((-t|--type-support) IDENTIFIER(==VERSION)?)* ((-I|--include-path) PATH)* PATH+
+```
+
+All positional arguments are paths to interface definition files.
+If no language / type support is specified, all type / type support generators available are used.
+If no type / type support generator is found, the command fails.
+If no output file path is provided, the generated build specification is sent to standard output.
+
+##### Implementation considerations
+
+All build logic in existing generator and type support packages is CMake code, and thus a port is necessary. 
+Using Python would ensure tooling interoperability and their supporting libraries.
+Additionally, [`setuptools` entrypoints](https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html#advertising-behavior) can be leveraged as plugin system.
+
+Since common build specifications are eminently declarative, most configuration must be resolved prior to generation e.g. for which OS to build.
+This includes dependencies between generated outputs e.g. type support code may depend on type representation code to build, and thus a build specification for the former must ensure the presence of the latter by conditionally generating it if not present.
+
+#### Meta build tool
+
+##### Description
+
+Develop an extensible, standardized tool that can translate [common build specifications](#a-common-build-specification) into an actual build system or build system generator code.
+Support for new build systems and build system generators can be added by external packages via a plugin system.
+
+##### Specification
+
+```sh
+ament_meta_build ((-o|--output-file) PATH)? ((-b|--build-system) IDENTIFIER)? PATH?
+```
+
+Its only positional argument is a path to a common build specification file.
```

Comment:
Good point. I wrote this under the assumption that the build system or build system generator this tool generates code for already has a default path resolution logic (like CMake with its binary and source trees). But perhaps that's not enough and this tool could be given that information. WDYT?

Reply:
That is, the CMake output plugin to `ament_meta_build` would assume relative paths in `Sources` come from `CMAKE_CURRENT_SOURCE_DIR`, things from `Artifacts` are relative to `CMAKE_CURRENT_BINARY_DIR`, and installed things from `Distribution` are relative to `CMAKE_INSTALL_PREFIX`?

Reply:
Only one thing to correct:

> That is, the CMake output plugin to `ament_meta_build` would assume relative paths in `Sources` come from **`CMAKE_CURRENT_SOURCE_DIR` or `CMAKE_CURRENT_BINARY_DIR`**, things from `Artifacts` are relative to `CMAKE_CURRENT_BINARY_DIR`, and installed things from `Distribution` are relative to `CMAKE_INSTALL_PREFIX`.

This is because a component's artifacts may be another component's sources. The lack of a distinction between source and build trees is because not all build system support this notion. Build specifications must make sure there's no ambiguity upon source lookup.